src/main/scala/mesosphere/util/state/zk/ZKStore.scala

```diff
@@ -90,22 +87,26 @@ class ZKStore(val client: ZkClient, rootNode: ZNode) extends PersistentStore {
     }
   }
 
-  private[this] def createPathBlocking(path: ZNode): ZNode = {
-    def createParent(node: ZNode): ZNode = {
-      val exists = Await.result(node.exists().asScala.map(_ => true)
-        .recover { case ex: NoNodeException => false }
-        .recover(exceptionTransform("Can not query for exists")), Duration.Inf)
-
-      if (!exists) {
-        createParent(node.parent)
-        Await.result(node.create().asScala
-          .recover { case ex: NodeExistsException => node }
-          .recover(exceptionTransform("Can not create")), Duration.Inf)
+  private[this] def createPath(path: ZNode): Future[ZNode] = {
+    def nodeExists(node: ZNode): Future[Boolean] = node.exists().asScala
+      .map(_ => true)
+      .recover { case ex: NoNodeException => false }
+      .recover(exceptionTransform("Can not query for exists"))
+
+    def createNode(node: ZNode): Future[ZNode] = node.create().asScala
+      .recover { case ex: NodeExistsException => node }
+      .recover(exceptionTransform("Can not create"))
+
+    def createParent(node: ZNode): Future[ZNode] = {
```

Comment:
Better use `createParentsAndNode` or something like that. This way it is misleading.